This Month in Linebender, September 2024 (#70)
linebender/linebender.github.io: content/blog/2024-10-04-tmil-09/index.md

<img style="height: auto; margin: 0 auto" src="../../linebender.svg" alt="Linebender logo; smoothly stroked overlapping lowercase letters, an l and a b." height="200" width = "200">

## Xilem

Xilem is our flagship GUI project, inspired by SwiftUI.
It lets you build user interfaces declaratively by composing lightweight views together, and will diff them to provide minimal updates to a retained layer.

- HTTP cats example ([xilem#571][]).
- `lens`, a shorthand for a common component pattern ([xilem#587][]).
- We added Views for some existing and new widgets:
  - Grid layout ([xilem#570][]).
  - Images ([xilem#571][]).

<!-- TODO: Xilem Web seems to not have any especially relevant updates? -->

<figure>

<img style="height: auto" src="http-cats.png" alt="A list of HTTP status codes, with 'Select' buttons. HTTP code 418 'I'm a teapot' is selected, with a picture of a kitten hiding in a teapot. height="962" width = "600">

<figcaption>

The HTTP cats example.

</figcaption>
</figure>

## Masonry

Masonry is the widget system used by Xilem.

> It provides a non-opinionated retained widget tree on top of Winit, designed as a base layer for high-level GUI frameworks.

Author: ```suggestion
It provides a non-opinionated retained widget tree, designed as a base layer for high-level GUI frameworks.
```

Our ongoing pass specification project has been largely implemented.
That work is now in a wrapping-up stage, which has allowed Olivier to work on improving the codebase's quality in general.
Some highlights of work from September:

- [xilem#515][]: Removes unimplemented functionality from our text handling code.
- [xilem#570][]: An implementation of a grid layout, by Jared O'Connell.
- [xilem#550][]: Restores scroll requests, as used by text inputs.
- [xilem#565][]: Allows widgets to implement more efficient pointer position dispatching, by new Linebender member Tom Churchman.
- [xilem#580][]: Adds screen reader support for the textbox widget.

<!-- No grid example, it's not really visual. -->

Some ongoing work includes:

- [xilem#615][]: Detailed accessibility information for text input and prose widgets.
- [xilem#606][]: Custom client side decorations for Wayland, by Marco Melorio.
- [xilem#605][]: Greater flexibility in the layout of the `Image` widget, by failingprovince.
- [xilem#591][]: An SVG-like widget for absolute positioning, by Muhammad Ragib Hasin.

## Vello

Vello is our GPU vector renderer.

> It can draw large 2D scenes with high performance, using GPU compute shadersfor most of the work.

Author: ```suggestion
It can draw large 2D scenes with high performance, using GPU compute shaders for most of the work.
```

> Progress towards the upcoming 0.3.0 release can be tracked in [the milestone](https://github.com/linebender/vello/milestone/2).

Author: ```suggestion
```

Since we have now released version 0.3.0, it's probably more likely to confuse if we include this.
(I also don't think we should post this update on reddit, given how much things have changed in the past two weeks)
